File: datastruct/map/BSTMap.java
public class BSTMap<K extends Comparable <K>, V> implements Map<K,V> {

    private class Node{
        public K key;
        public V val;
        public Node left;
        public Node right;

        public Node(K key, V val) {
            this.key = key;
            this.val = val;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;
    private int size;

    public BSTMap(){
        root = null;
        size = 0;
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }

    @Override
    public int getSize() {
        return size;
    }

    // 向二分搜索树中添加元素（key，value）
    @Override
    public void add(K key, V value) {
        root = add(root, key, value);
    }

    // 向以 node 为根的二分搜索树添加元素（key，value），递归实现
    // 返回添加元素后的根结点
    private Node add(Node node, K key, V value) {

        if(node == null){
            size++;
            return new Node(key, value);
        }

        if(key.compareTo(node.key) > 0 )
            node.right = add(node.right, key, value);
        else if(key.compareTo(node.key) < 0)
            node.left = add(node.left, key, value);
        else // key.compareTo(node.key) == 0
            node.val = value;

        return node;
    }

    // 返回以 node 为根结点的二分搜索树中关键字为 key 的结点
    private Node getNode(Node node, K key){
        if(node == null)
            return null;

        if(key.compareTo(node.key) < 0)
            return getNode(node.left, key);
        else if(key.compareTo(node.key) > 0)
            return getNode(node.right, key);
        else // key.compareTo(node.key) == 0
            return node;
    }

    @Override
    public boolean contain(K key) {
        return getNode(root, key) != null;
    }

    @Override
    public V get(K key) {
        Node node = getNode(root, key);
        return node == null ? null : node.val;
    }

    @Override
    public void set(K key, V value) {
        Node node = getNode(root, key);

        if(node == null)
            throw new IllegalArgumentException(key + " doesn't exist!");

        node.val = value;
    }

    // 返回以 node 为根结点的二分搜索树中的最小节点
    private Node min(Node node){
        if(node == null)
            throw new IllegalArgumentException("Tree is empty!");

        while(node.left != null)
            node = node.left;

        return node;
    }

    // 从以 node 为根结点的二分搜索树中删除 key 所在的结点
    // 返回删除元素后的根结点
    private Node removeMin(Node node, K key){
        Node ret = min(node);
        while(node.left != ret)
            node = node.left;
        node.left = ret.right;
        size--;
        return ret;
    }

    @Override
    public V remove(K key) {
        Node node = getNode(root, key);
        if(node == null)
            return null;
        root = remove(root, key);

        return node.val;
    }

    // 从以 node 为根结点的二分搜索树中删除 key 所在的结点
    // 返回删除结点后的根结点
    private Node remove(Node node, K key) {

        if(key.compareTo(node.key) < 0){
            node.left = remove(node.left, key);
        }
        else if(key.compareTo(node.key) > 0){
            node.right = remove(node.right, key);
        }
        else{ // key.compareTo(node.key) == 0
            if(node.left == null){
                Node ret = node.right;
                node.right = null;
                size--;
                return ret;
            }
            else if(node.right == null){
                Node ret = node.left;
                node.left = null;
                size--;
                return ret;
            }
            else{
                Node ret = min(node.right);
                ret.right = removeMin(node.right, key);
                ret.left = node.left;
                node.left = null;
                node.right = null;
                return ret;
            }
        }
        return node;
    }
}

File: datastruct/map/BSTMap.py
from Map import Map

class Node(object):

    def __init__(self, key=None, value=None, left=None, right=None):
        self.key = key
        self.value = value
        self.left = left
        self.right = right
    
    def __str__(self):
        return str(self.key) + ': ' + str(self.value)

class BSTMap(Map):

    def __init__(self):
        self.__size = 0
        self.__root = None
    
    def is_empty(self):
        return self.__size == 0
    
    def get_size(self):
        return self.__size

    def add(self, key, value):
        self.__root = self.__add(self.__root, key, value)

    
    def __add(self, node: Node, key, value):

        if node is None:
            self.__size += 1
            return Node(key, value)
        
        if key < node.key:
            node.left = self.__add(node.left, key, value)
        elif key > node.key:
            node.right = self.__add(node.right, key, value)
        else:
            node.value = value
        
        return node

    def get_node(self, key):
        if key is None:
            return None
        return self.__get_node(self.__root, key)
    
    def __get_node(self, node: Node, key):

        if node is None:
            return None

        if key < node.key:
            return self.__get_node(node.left, key)
        elif key > node.key:
            return self.__get_node(node.right, key)
        else:
            return node

    def contains(self, key):
        node = self.get_node(key)

        if node is None:
            return False
        else:
            return True
    
    def get(self, key):
        node = self.get_node(key)

        if node is None:
            return None
        else:
            return node.value

    def set(self, key, value):
        node = self.get_node(key)

        if node is None:
            raise IndexError('%s does not exist!' % key)

        node.value = value
    
    def __min(self, node: Node):

        if node is None:
            raise IndexError('Tree is empty!')

        if node.left is not None:
            return self.__min(node.left)
        
        return node

    def remove(self, key):
        ret = self.get_node(key)
        if ret is not None:
            ret = ret.value
            self.__root = self.__remove(self.__root, key)
        return ret

    def __remove(self, node: Node, key):

        if node is None:
            raise IndexError('Tree is empty!')

        if key < node.key:
            node.left = self.__remove(node.left, key)
        elif key > node.key:
            node.right = self.__remove(node.right, key)
        else: # key == node.key
            if node.left is None:
                self.__size -= 1
                ret = node.right
                node.right = None
                return ret
            elif node.right is None:
                self.__size -= 1
                ret = node.left
                node.left = None
                return ret
            else:
                ret = self.__min(node.right)
                ret.right = self.__remove(node.right, ret.key)
                ret.left = node.left
                node.right = node.left = None
                return ret

        return node
